### PR TITLE
PBKDF2 updates to conform to SP800-132

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: crypto/kdf/scrypt.c [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017-2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017-2019 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (paulidale) [inline crypto/kdf/scrypt.c]: Revert this change please.

File: doc/man7/EVP_KDF_PBKDF2.pod [hunk 2]
```diff
@@ -34,9 +34,21 @@ The supported controls are:
...
 
-B<iter> is the iteration count and its value should be greater than or equal to
-1. RFC 2898 suggests an iteration count of at least 1000.  The default value is
-2048.  Any B<iter> less than 1 is treated as a single iteration.
+=item B<EVP_KDF_CTRL_SET_PBKDF2_PKCS5_MODE>
+
+This control expects one argument: C<int checks>
+
+This control can be used to turn off the compliance checks. This can be used for
+PKCS5 related test vectors and scrypt which do not comply to SP800-132.
```
Comment (paulidale) [inline doc/man7/EVP_KDF_PBKDF2.pod]: "comply with" or "conform to"

File: doc/man7/EVP_KDF_PBKDF2.pod [hunk 3]
```diff
@@ -34,9 +34,21 @@ The supported controls are:
...
+The compliance checks performed are:
+A minimum iteration count of 1000 and length of the salt shall be at least
```
Comment (paulidale) [inline doc/man7/EVP_KDF_PBKDF2.pod]: What about _... and the length of the salt being at least_ instead?

File: test/evp_kdf_test.c [hunk 4]
```diff
@@ -70,24 +70,55 @@ static int test_kdf_pbkdf2(void)
...
+                                      "saltSALTsaltSALTsaltSALTsaltSALTsalt",
+                                      (size_t)36), 0)
+          && TEST_int_gt(EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_SET_ITER, 4096), 0)
           && TEST_int_gt(EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_SET_MD, EVP_sha256()),
                          0)
           && TEST_int_gt(EVP_KDF_derive(kctx, out, sizeof(out)), 0)
-          && TEST_mem_eq(out, sizeof(out), expected, sizeof(expected));
+          && TEST_mem_eq(out, sizeof(out), expected, sizeof(expected))
+          /* A key length that is too small should fail */
+          && TEST_int_eq(EVP_KDF_derive(kctx, out, 112/8 - 1), 0)
```
Comment (paulidale) [inline test/evp_kdf_test.c]: Spaces are the `/`.

File: test/evp_kdf_test.c [hunk 5]
```diff
@@ -70,24 +70,55 @@ static int test_kdf_pbkdf2(void)
 {
     int ret;
     EVP_KDF_CTX *kctx;
-    unsigned char out[32];
+    unsigned char out[25];
```
Comment (paulidale) [inline test/evp_kdf_test.c]: Would this be better declared later using `sizeof(expected)` to avoid magic numbers?

File: test/evp_kdf_test.c (hunk 4 above)
Comment (slontis) [reply test/evp_kdf_test.c]: I dont know what you mean

Comment (paulidale) [reply test/evp_kdf_test.c]: `128 / 8` instead of `128/8`

File: test/evp_kdf_test.c (hunk 5 above)
Comment (slontis) [reply test/evp_kdf_test.c]: I dont think this makes much difference. the expected[] still needs to be a fixed size. The KAT that it is taken from has a fixed length, so the current way ensures the outlength is correct.. even if the expected array was incorrect and got zero padded. Either way the KAT would fail and the issue would be resolved.

File: doc/man7/EVP_KDF_PBKDF2.pod [hunk 6]
```diff
@@ -34,9 +34,21 @@ The supported controls are:
...
 
-B<iter> is the iteration count and its value should be greater than or equal to
-1. RFC 2898 suggests an iteration count of at least 1000.  The default value is
-2048.  Any B<iter> less than 1 is treated as a single iteration.
+=item B<EVP_KDF_CTRL_SET_PBKDF2_PKCS5_MODE>
+
+This control expects one argument: C<int checks>
+
+This control can be used to turn off the compliance checks. This can be used for
+PKCS5 related test vectors and scrypt which do not comply with SP800-132.
```
Comment (kaduk) [inline doc/man7/EVP_KDF_PBKDF2.pod]: I think that to retain feature compatibility (with strange/semi-broken deployments) Kerberos users of PBKDF2 will also need to set this option.  So I think rewording this to not sound like an exhaustive list would be wise.

Comment (slontis) [reply doc/man7/EVP_KDF_PBKDF2.pod]: So you just want 
>This control can be used to turn off the compliance checks. This can be used for
>PKCS5 related test vectors and scrypt which do not comply with SP800-132.
reduced to..
This control can be used to turn off SP800-132 compliance checks.

We could also only enable this by default in fips mode? (But they do seem to be fairly valid checks though). We could also just not do the iteration check as well (as its only recommended - the salt however is a 'shall be statement')

Comment (kaduk) [reply doc/man7/EVP_KDF_PBKDF2.pod]: That would work, as would "can be used for cases such as [...]" or "This is needed in some cases, including but not limited to, [...]".

**Comment by slontis** [comment]
Updated commit message and rebased.

**Comment by mspncp** [comment]
Would you mind using present tense in the commit message of 1827b7119195? See https://github.com/openssl/openssl/pull/8763#issuecomment-496244670 for an explanation.

```
PBKDF2: conform to SP800-132

The existing code used PKCS5 specifications.
SP800-132 adds the following additional constraints:
  - Check the range of the key length.
  - Minimum iteration count (1000 recommended).
  - Randomly-generated portion of the salt shall be at least 128 bits.
There are cases where the constraints will cause errors (in scrypt, and
some PKCS5 related test vectors) so a ctrl "pkcs5" is added to disable
these additional constraints.
```

**Comment by slontis** [comment]
Updated the commit message and rebased. (This present tense stuff hurts my brain.)

**Comment by mspncp** [comment]
> Updated the commit message and rebased. (This present tense stuff hurts my brain.)

Sorry that it hit you so often. I did not mean to stalk you ;-)

**Comment by mattcaswell** [comment]
Looks like you've picked up @t8m's commit in this PR (I guess you rebased on top of github master in the window while 6acf605 was there, before it got force pushed away - see the discussion in #9015). Can you rebase this?

File: doc/man7/EVP_KDF_PBKDF2.pod [hunk 7]
```diff
@@ -34,9 +34,22 @@ The supported controls are:
...
+This control expects one argument: C<int mode>
+
+This control can be used to turn off SP800-132 compliance checks. This is needed
+in some cases, including but not limited to, PKCS5 related test vectors and
+scrypt.
+The compliance checks performed are:
+A minimum iteration count of 1000 and length of the salt being at least
+128 bits.
+
+The default value is 0. Use 1 to disable the checks.
```
Comment (mattcaswell) [inline doc/man7/EVP_KDF_PBKDF2.pod]: So, this sounds like having the default like this is a breaking change?

File: crypto/kdf/pbkdf2.c [hunk 8]
```diff
@@ -209,7 +231,17 @@ static int pkcs5_pbkdf2_alg(const char *pass, size_t passlen,
     HMAC_CTX *hctx_tpl = NULL, *hctx = NULL;
 
     mdlen = EVP_MD_size(digest);
-    if (mdlen < 0)
+    if (mdlen <= 0)
+        return 0;
+
+    /* Check that 112 bits <= keylen < (2^32 - 1) * mdlen */
+    if (((keylen * 8) < KDF_PBKDF2_MIN_KEY_LEN_BITS)
+        || ((keylen / mdlen) >= KDF_PBKDF2_MAX_KEY_LEN_DIGEST_RATIO))
```
Comment (mattcaswell) [inline crypto/kdf/pbkdf2.c]: Should we be also checking the value of `range_checks` here?

File: doc/man7/EVP_KDF_PBKDF2.pod [hunk 9]
```diff
@@ -34,9 +34,22 @@ The supported controls are:
 
 These controls work as described in L<EVP_KDF_CTX(3)/CONTROLS>.
 
-B<iter> is the iteration count and its value should be greater than or equal to
-1. RFC 2898 suggests an iteration count of at least 1000.  The default value is
-2048.  Any B<iter> less than 1 is treated as a single iteration.
```
Comment (mattcaswell) [inline doc/man7/EVP_KDF_PBKDF2.pod]: We seem to have lost the information that the default number of iterations is 2048.

File: crypto/kdf/pbkdf2.c [hunk 10]
```diff
@@ -105,11 +117,14 @@ static int kdf_pbkdf2_ctrl(EVP_KDF_IMPL *impl, int cmd, va_list args)
...
         len = va_arg(args, size_t);
+        if (impl->range_checks != 0 && len < KDF_PBKDF2_MIN_SALT_LEN)
+            return 0;
         return pbkdf2_set_membuf(&impl->salt, &impl->salt_len, p, len);
 
     case EVP_KDF_CTRL_SET_ITER:
         iter = va_arg(args, int);
-        if (iter < 1)
+        min_iter = impl->range_checks != 0 ? KDF_PBKDF2_MIN_ITERATIONS : 1;
+        if (iter < min_iter)
```
Comment (mattcaswell) [inline crypto/kdf/pbkdf2.c]: Should we be adding errors to the queue in these failure scenarios?

**Comment by slontis** [comment]
Ahh that would explain the merge commit I had earlier in the day.. sigh.

File: crypto/kdf/pbkdf2.c (hunk 8 above)
Comment (slontis) [reply crypto/kdf/pbkdf2.c]: I did ponder it - But I think not..
If it goes above that range the counter overflows which is very bad.

Comment (mattcaswell) [reply crypto/kdf/pbkdf2.c]: > If it goes above that range the counter overflows which is very bad.

Perhaps a comment stating that would be helpful

Is that also true for the `KDF_PBKDF2_MIN_KEY_LEN_BITS` check?

Comment (slontis) [reply crypto/kdf/pbkdf2.c]: Not sure  - I assume that is similar to the salt check so maybe that one goes in the check - in that case I might rename the range_checks variable :)

File: doc/man7/EVP_KDF_PBKDF2.pod (hunk 7 above)
Comment (slontis) [reply doc/man7/EVP_KDF_PBKDF2.pod]: enc uses the PKCS5_PBKDF2_HMAC() method which sets this control to be 1, as does scrypt.

File: crypto/kdf/pbkdf2.c (hunk 10 above)
Comment (slontis) [reply crypto/kdf/pbkdf2.c]: Added

File: crypto/kdf/pbkdf2.c (hunk 8 above)
Comment (slontis) [reply crypto/kdf/pbkdf2.c]: Comment added and the lower bound check has been moved.

File: doc/man7/EVP_KDF_PBKDF2.pod (hunk 7 above)
Comment (slontis) [reply doc/man7/EVP_KDF_PBKDF2.pod]: Added note to CHANGES

**Comment by slontis** [comment]
ping @mattcaswell

**Comment by mattcaswell** [comment]
This looks ok, but since this is a breaking change I'd like a wider viewpoint from @openssl/omc.

**Comment by slontis** [comment]
Note that the minimum iteration count is only recommended (it is not a SHALL) so that particular test could be relaxed (but it is probably a better idea to still force it to at least the recommended value)

**Comment by levitte** [comment]
Whether this is a breaking change or not depends on your point of view. EVP_KDF is an unreleased API, so there's nothing that can be called a breaking change there. However, for KDFs implemented via EVP_PKEY, this is effectively a breaking change.

To resolve that issue, the use via EVP_PKEY could be relaxed so as not to pose a problem.

**Comment by kroeckx** [comment]
I think we should support reading old data from before this change.

**Comment by slontis** [comment]
> To resolve that issue, the use via EVP_PKEY could be relaxed so as not to pose a problem.

Ok so that would be a fairly ugly check. The best solution would probably be to have it default to how it used to work, but in FIPS_MODE it does the extra checks by default. The flag can still be enabled or disabled in either provider.
Does that sound like a better solution?

**Comment by slontis** [comment]
Updated the default value to disable the checks by default in the default provider.

**Comment by mattcaswell** [comment]
> Updated the default value to disable the checks by default in the default provider.

Sounds like a reasonable approach.

**Comment by levitte** [comment]
I agree with @mattcaswell. Good thinking!

**Comment by slontis** [comment]
ping @mattcaswell or @levitte : Code was updated to set default check for FIPS_MODE only.

**Comment by slontis** [comment]
rebased to fix CHANGES conflict.
